src/main/java/org/dependencytrack/persistence/VulnerabilityQueryManager.java

```diff
@@ -524,6 +526,23 @@ public List<Project> getProjects(final Vulnerability vulnerability, final Set<St
         return getObjectsById(Project.class, affectedProjectIds, fetchGroups);
     }
 
+    /**
+     * Returns a List of Projects affected by a specific vulnerability.
+     * @param vulnerability the vulnerability to query on
+     * @return a List of AffectedProjects
+     */
+    public List<AffectedProject> getAffectedProjects(Vulnerability vulnerability) {
+        return this.getProjects(vulnerability, Set.of(Project.FetchGroup.IDENTIFIERS.name())).stream()
```

Comment:
OK, figured out what's missing to make the test pass.

In the code the original PR is based on, we query all components affected by the given `vulnerability`, and then iterate over those to figure out which (unique) projects are affected. That was horribly inefficient, so in Hyades we replaced it with a query that directly fetches the affected projects.

`VulnerabilityResourceTest#getAffectedProjectTest` is currently failing because the `$.affectedComponentUuids` JSON array is empty.

Because the original PR iterated over all components, it had access to the UUIDs of all affected components: https://github.com/DependencyTrack/dependency-track/pull/3285/files#diff-dd29b05734d1b93668ceb9c364622122eee874786f6ff8c0c13275349dac6a64R500

That is no longer the case. We now have a list of affected projects, and need to figure out the UUIDs of the components within them that are affected. This could be done in *one* additional query.

Here's an alternative implementation of `getAffectedProjects` that makes the test pass. Added some comments for explanation, but I'd encourage you to review and adjust as you see fit.

```java
public List<AffectedProject> getAffectedProjects(Vulnerability vulnerability) {
    // Fetch all projects affected by the given vulnerability.
    // Group them by their UUID to ease additional enrichment.
    final Map<String, AffectedProject> affectedProjects = getProjects(vulnerability, Set.of(Project.FetchGroup.IDENTIFIERS.name())).stream()
        .map(project -> new AffectedProject(
            project.getUuid(),
            project.getDirectDependencies() != null,
            project.getName(),
            project.getVersion(),
            null
        ))
        .collect(Collectors.toMap(affectedProject -> affectedProject.getUuid().toString(), Function.identity()));

    // Fetch UUIDs of components that are part of the affected projects,
    // and affected by the given vulnerability. Return both the component's UUID,
    // but also the UUID of the project it belongs to, so we can correlate.
    final Query<Component> query = pm.newQuery(Component.class);
    query.setFilter(":projectUuids.contains(project.uuid) && vulnerabilities.contains(:vulnerability)");
    query.setParameters(affectedProjects.keySet(), vulnerability);
    query.setResult("uuid, project.uuid");
    final List<Object[]> resultRows;
    try {
        resultRows = List.copyOf(query.executeResultList(Object[].class));
    } finally {
        query.closeAll();
    }

    // "Enrich" affectedProjects with component UUIDs.
    for (final Object[] resultRow : resultRows) {
        final String componentUuid = String.valueOf(resultRow[0]);
        final String projectUuid = String.valueOf(resultRow[1]);
        affectedProjects.get(projectUuid).getAffectedComponentUuids().add(UUID.fromString(componentUuid));
    }

    return new ArrayList<>(affectedProjects.values());
}
```

In the future we might want to convert all this to an SQL query. Most of this can be achieved with aggregations in SQL, but JDO queries are not as powerful.